design/kanister-prometheus-integration.md

```diff
@@ -0,0 +1,290 @@
+	return &Controller{
+		config:  c,
+		metrics: newMetrics(prometheus.DefaultRegistry), // this helper method call will be made during init
+	}
+}
+```
+
+
+controller/metrics.go
+
+```golang
+const (
+    ACTION_SET_COUNTER_VEC_LABEL_RES = "resolution"
+    ACTION_SET_COUNTER_VEC_LABEL_OP_TYPE = "operation_type"
+)
+
+type metrics struct {
+    ActionSetCounterVec *prometheus.CounterVec
+}
+
+// helper method to construct the correct "LabelHeaders":"LabelValues" mapping
+// to ensure type safety
+func getActionSetCounterVecLabels() []kanistermetrics.BoundedLabels {
+    bl := make([]kanistermetrics.BoundedLabel, 2)
+	  bl[0] = kanistermetrics.BoundedLabel{LabelName: ACTION_SET_COUNTER_VEC_LABEL_RES,
+		LabelValues: []string{"success", "failure"}}
+    bl[1] = kanistermetrics.BoundedLabel{LabelName:
+    ACTION_SET_COUNTER_VEC_LABEL_BLUEPRINT, 
+    LabelValues: []string{"backup", "restore"}}
+    return bl
+}
+
+
+// helper method to construct the labels correctly. 
+func constructActionSetCounterVecLabels(operation_type string, resolution string) prometheus.Labels {
+	return prometheus.Labels{ACTION_SET_COUNTER_VEC_LABEL_OP_TYPE: operation_type, 
+  ACTION_SET_COUNTER_VEC_LABEL_RES: resolution}
+}
+
+// helper method to create a Metrics interface
+func newMetrics(gatherer prometheus.Gatherer) *metrics {
+    actionSetCounterOpts := prometheus.CounterOpts{
+        Name: "action_set_resolutions_total",
+        Help: "Total number of action set resolutions",
+    }
+    actionSetCounterVec := kanistermetrics.InitCounterVec(gatherer,
+		actionSetCounterOpts, getActionSetCounterVecLabels())
+    return &metrics{ActionSetCounterVec: actionSetCounterVec}
+}
+```
+
+The below example will show how the above created ActionSetCounterVec 
+will be incremented in a method:
+
+```golang
+func (c *Controller) handleActionSet(ctx context.Context) {
+  c.metrics.ActionSetCounterVec.With(constructActionSetCounterVecLabels("backup", "success")).Inc()
+}
+```
+
+Alternatively, one can also directly call the Prometheus API with 
+positional arguments:
+
+```golang
+func (c *Controller) handleActionSet(ctx context.Context) {
+  c.metrics.ActionSetCounterVec.WithLabelValues("backup", "success").Inc()
+}
+```
+
+
+## Testing
+
+1. The testing will include manual testing of whether the metrics added are 
+   successfully getting exported to Kanister. 
+
+2. The interfaces listed above in the metrics package, apart from InitCounterVec 
+   and generateLabelCombinations, will not be unit-tested, since they would be 
+   testing the behavior of the Prometheus API itself, which breaks the chain 
+   of trust principle with dependencies in unit testing.  
+   InitCounterVec will be unit tested using the test_util 
+   package: https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/testutil 
+   in Prometheus.
+
+3. Integration tests will be added for code that exports new metrics, 
+   to ensure that the behavior of exporting metrics is correct. 
```

Comment:
```suggestion
1. The testing will include manual testing of whether the metrics added are
   successfully getting exported to Kanister.

2. The interfaces listed above in the metrics package, apart from
   `InitCounterVec` and `generateLabelCombinations`, will not be unit-tested,
   since they would be testing the behavior of the Prometheus API itself, which
   breaks the chain of trust principle with dependencies in unit testing.
   `InitCounterVec` will be unit tested using the
   [test_util](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/testutil)
   package in Prometheus.

3. Integration tests will be added for code that exports new metrics, to ensure
   that the behavior of exporting metrics is correct. 
```